File: app/src/main/java/de/westnordost/streetcomplete/quests/bridge_structure/AddBridgeStructure.java
package de.westnordost.streetcomplete.quests.bridge_structure;

import android.os.Bundle;
import android.support.annotation.NonNull;

import java.util.List;
import java.util.Map;

import javax.inject.Inject;

import de.westnordost.streetcomplete.R;
import de.westnordost.streetcomplete.data.osm.SimpleOverpassQuestType;
import de.westnordost.streetcomplete.data.osm.changes.StringMapChangesBuilder;
import de.westnordost.streetcomplete.data.osm.download.OverpassMapDataDao;
import de.westnordost.streetcomplete.quests.AbstractQuestAnswerFragment;

public class AddBridgeStructure extends SimpleOverpassQuestType
{
	@Inject public AddBridgeStructure(OverpassMapDataDao overpassServer) { super(overpassServer); }

	@Override protected String getTagFilters()
	{
		return "ways with man_made=bridge and !bridge:structure";
	}

	@Override public AbstractQuestAnswerFragment createForm() { return new AddBridgeStructureForm(); }

	@Override public void applyAnswerTo(Bundle answer, StringMapChangesBuilder changes)
	{
		List<String> values = answer.getStringArrayList(AddBridgeStructureForm.OSM_VALUES);
		if(values != null  && values.size() == 1)
		{
			changes.add("bridge:structure", values.get(0));
		}
	}

	@Override public int getTitle(@NonNull Map<String, String> tags) { return R.string.quest_bridge_structure_title; }
	@Override public int getIcon() { return R.drawable.ic_quest_bridge; }
	@Override public String getCommitMessage() { return "Add bridge structures"; }
}

File: app/src/main/java/de/westnordost/streetcomplete/quests/bridge_structure/AddBridgeStructure.kt
package de.westnordost.streetcomplete.quests.bridge_structure

import de.westnordost.streetcomplete.R
import de.westnordost.streetcomplete.data.osm.geometry.ElementGeometry
import de.westnordost.streetcomplete.data.osm.osmquests.OsmFilterQuestType
import de.westnordost.streetcomplete.data.user.achievements.EditTypeAchievement.BUILDING
import de.westnordost.streetcomplete.osm.Tags

class AddBridgeStructure : OsmFilterQuestType<BridgeStructure>() {

    override val elementFilter = """
        ways with
          man_made = bridge
          and !bridge:structure
          and !bridge:movable
          and (!indoor or indoor = no)
    """
    override val changesetComment = "Specify bridge structures"
    override val wikiLink = "Key:bridge:structure"
    override val icon = R.drawable.ic_quest_bridge
    override val achievements = listOf(BUILDING)

    override fun getTitle(tags: Map<String, String>) = R.string.quest_bridge_structure_title

    override fun createForm() = AddBridgeStructureForm()

    override fun applyAnswerTo(answer: BridgeStructure, tags: Tags, geometry: ElementGeometry, timestampEdited: Long) {
        tags["bridge:structure"] = answer.osmValue
    }
}
